File: Neetcode-150-js/Linked List/copy-list-with-random-pointer.js

// https://leetcode.com/problems/copy-list-with-random-pointer/

/**
 * // Definition for a _Node.
 * function _Node(val, next, random) {
 *    this.val = val;
 *    this.next = next;
 *    this.random = random;
 * };
 */

/**
 * @param {_Node} head
 * @return {_Node}
 */
var copyRandomList = function(head) {

    // base case
    if(!head) return head;

    let temp = head;
    while(temp){
        let copyNode = new Node(temp.val);
        let nextNode = temp.next;
        temp.next = copyNode;
        copyNode.next = nextNode;
        temp = temp.next.next;
    }

    temp = head;
    while(temp){
        if(temp.random) temp.next.random = temp.random.next;
        temp = temp.next.next;
    }

    let copyHead = new Node(null, null, null);
    let copyTemp = copyHead;
    temp = head;
    while(temp){
        let nextNode = temp.next.next;
        copyTemp.next = temp.next;
        copyTemp = copyTemp.next
        temp.next = nextNode;
        temp = temp.next;
    }

    return copyHead.next;
};

File: Striver Important Interview Questions/Day 7 - Linked List & Arrays/copy-list-with-random-pointer.cpp

// https://leetcode.com/problems/copy-list-with-random-pointer/

/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        if(!head) return NULL;

        Node*nextNode = NULL;
        Node*curr = head;

        // creating duplicate nodes
        while(curr){
            nextNode = curr->next;
            Node*copyNode = new Node(curr->val);
            curr->next = copyNode;
            copyNode->next = nextNode;
            curr = nextNode;
        }

        // assiging random pointers to the created nodes
        curr = head;
        while(curr){
            if(curr->random) curr->next->random = curr->random->next;
            curr = curr->next->next;
        }

        // segregating both the lists
        Node*copy = new Node(-1);
        Node*temp = copy;
        curr = head;
        nextNode = NULL;
        while(curr){
            nextNode = curr->next->next;
            temp->next = curr->next;
            temp = temp->next;
            curr->next = nextNode;
            curr = curr->next;
        }
        
        return copy->next;
    }
};